### ISSUE TITLE
Cannot indicate dependencies to useSprings

### ISSUE THREAD

**Issue by bebbi** [react-spring/react-spring#704]
## 🐛 Bug Report

In [this sandbox](https://codesandbox.io/s/dazzling-varahamihira-t1l4k), I'm re-using the `DraggableList` example and integrating it with external state that is passed to the component as a prop.

When I sort items, or new items are added, an updated `items` prop is sent to the component but it fails to maintain the sort order correctly. Only after another click on a sortable item will the sort order be accurate.

## To Reproduce

In the sandbox, move any non-separator item, wait a second to see them become mis-ordered, then click into a yellow item to have them ordered correctly.

## Expected behavior

2x1 seconds after the updated items array is sent to the component, there's a `setSpring` call which appears to create a correct spring setup but is not rendered (line 76).

This doesn't get rendered when forcing a react render either, otherwise this could have been a work-around.

I would expect that line to immediately trigger a re-render.

After trying for hours to integrate this component with external state I'm thinking this might be a bug.

## Link to repro (highly encouraged)

See codesandbox link above.

## Environment

- `react-spring` v8.0.19
- `react` v16.8.6

**Comment by bebbi** [comment]
The issue is that a setter of the form

`const [springs, setSprings] = useSprings(itemsLength, index => fn(dep1, dep2)`

doesn't trigger or even support a manual re-render by itself, i.e. when no `useGesture` is active.

When `dep1, dep2` change, this will not re-render the updated springs, and will neither when forcing a react re-render through an unrelated state change.

It will however re-render reset springs if `itemsLength` changes.

How can I indicate the dependencies for re-render to `useSpring`?

Possibly something possible like:

```
const [springs, setSprings] = useSprings(
  itemsLength,
  index => fn(dep1, dep2,
  [dep1, dep2]
)
```

**Comment by bebbi** [comment]
This [SO question](https://stackoverflow.com/questions/56347324/using-react-spring-usesprings-with-dynamic-items-array) looks related.

**Comment by drcmda** [comment]
i believe you can do: useSprings(length, [{...}, {...}, {...}]) to override props on render.

but dependencies are really needed and will simplify the api. it will land in v9, which you can already run in the next branch. you can check out the ongoing PR for details.

**Comment by bebbi** [comment]
That sounds great thanks!

When refreshing, I've been thinking about the possibilty to maintain ongoing movement when refreshing the springs: e.g. by adding an array of unique spring keys `[key1, key2, ..]` as a replacement for `itemLength`.
This would map existing, moving spring params to new definitions and thus keep ongoing transitions active if the key existed before.

Seems like that would give for awesome and user-friendly effects on adding new data into lists.
Does that sound far-fetched?

**Comment by bebbi** [comment]
In `v9.0.0-beta.22`, below doesn't render updated spring params when the heights array changes, but instead still updates only on a gesture. I can see that the code now has a `deps` param. Is this a bug to report?

```
  const [springs, setSprings] = useSprings(
    items.length,
    getSpringParams(order, heights),
    [items, order, heights]
  )
```

**Comment by aleclarson** [comment]
@bebbi Nah, that's not a bug. Only shallow values are looked at in the `deps` array. That's how `useEffect` works too.



> When building a dependency management, I've been thinking about the possibilty to maintain ongoing movement when refreshing the springs: e.g. by adding an array of unique spring keys `[key1, key2, ..]` as a replacement for `itemLength`.
> This would map existing, moving spring params to new definitions and thus keep ongoing transitions active if the key existed before.

You should open a new issue for this proposal. And please provide some pseudocode so we can more easily visualize what you're talking about.

**Comment by bebbi** [comment]
From a not too informed look at the [code](https://github.com/react-spring/react-spring/blob/fb03c2defc1cab1a3a5e6792359affee79e85b1d/packages/core/src/useSprings.js#L75), could it be that the `useEffect` runs but skips the relevant parts because `hasNewSprings` is always `false` unless again the item count changes, since it exclusively relates to the `length` arg (`hasNewSprings = length !== usePrev(length))?

**Comment by bebbi** [comment]
@aleclarson, above is with shallow inequality. Would a repro be useful?
Sorry for the off-topic comment. It felt not worth an issue when 9 is not out.

**Comment by bebbi** [comment]
Also, the way the [deps](https://github.com/react-spring/react-spring/blob/0b8011c5b4e7e135eb7324a691beba936ea06036/packages/core/src/useSprings.js#L85) param is used in #2a87dd, I'd expect that this will call the `useEffect` on every render, because when calling it as usual:

```
const springState = useSprings(length, params, [dep1, dep2])
```

the caller actually necessarily creates a new deps array on every render. Which in turn leads to the `useEffect` function inside `useSprings` as linked above to run on every render too, instead of only when deps change, because it only does shallow comparisons.
